src/server/tablet_snap.rs

```diff
@@ -101,70 +176,412 @@ fn io_type_from_raft_message(msg: &RaftMessage) -> Result<IoType> {
+    missing_ssts: Vec<String>,
+    stream: &mut (impl Stream<Item = Result<TabletSnapshotRequest>> + Unpin),
+    limiter: &Limiter,
+) -> Result<u64> {
+    let mut received_bytes: u64 = 0;
+    for name in missing_ssts {
+        let chunk = match stream.next().await {
+            Some(Ok(mut req)) if req.has_chunk() => req.take_chunk(),
+            res => return Err(protocol_error("chunk", res)),
+        };
+        if chunk.file_name != name {
+            return Err(protocol_error(&name, &chunk.file_name));
+        }
+        received_bytes += accept_one_file(path, chunk, stream, limiter).await?;
+    }
+    // Now receive other files.
+    loop {
+        let chunk = match stream.next().await {
+            None => {
+                File::open(path)?.sync_data()?;
+                return Ok(received_bytes);
             }
-            limiter.consume(off).await;
-            total_sent += off as u64;
-            let mut chunk = SnapshotChunk::default();
-            chunk.set_data(buffer);
+            Some(Ok(mut req)) if req.has_chunk() => req.take_chunk(),
+            res => return Err(protocol_error("chunk", res)),
+        };
+        if chunk.file_name.is_empty() {
+            return Err(protocol_error("file_name", &chunk.file_name));
+        }
+        received_bytes += accept_one_file(path, chunk, stream, limiter).await?;
+    }
+}
+
+async fn recv_snap_files<'a>(
+    snap_mgr: &'a TabletSnapManager,
+    cache_builder: impl SnapCacheBuilder,
+    mut stream: impl Stream<Item = Result<TabletSnapshotRequest>> + Unpin,
+    sink: &mut (impl Sink<(TabletSnapshotResponse, WriteFlags), Error = grpcio::Error> + Unpin),
+    limiter: Limiter,
+) -> Result<RecvTabletSnapContext<'a>> {
+    let head = stream
+        .next()
+        .await
+        .transpose()?
+        .ok_or_else(|| Error::Other("empty gRPC stream".into()))?;
+    let context = RecvTabletSnapContext::new(head, snap_mgr)?;
+    let _with_io_type = WithIoType::new(context.io_type);
+    let region_id = context.key.region_id;
+    let final_path = snap_mgr.final_recv_path(&context.key);
+    if final_path.exists() {
+        // The snapshot is received already, should wait for peer to apply. If the
+        // snapshot is corrupted, the peer should destroy it first then request again.
+        return Err(Error::Other(
+            format!("snapshot {} already exists", final_path.display()).into(),
+        ));
+    }
+    let path = snap_mgr.tmp_recv_path(&context.key);
+    info!("begin to receive tablet snapshot files"; "file" => %path.display(), "region_id" => region_id);
+    if path.exists() {
+        fs::remove_dir_all(&path)?;
+    }
+    let (reused, missing_ssts) = if context.use_cache {
+        if let Err(e) = cache_builder.build(region_id, &path) {
+            info!("not using cache"; "region_id" => region_id, "err" => ?e);
+            fs::create_dir_all(&path)?;
+        }
+        cleanup_cache(&path, &mut stream, sink, &limiter).await?
```

Comment:
So it's possible that build returns error but path is not empty?

Reply:
Not possible, but the sender side has chosen to start cache exchange, so still need to call the function to complete the protocol.